---
title: "Intro to Transcriptomics"
author: Eros Mendoza
date: 07/17/22
updated: 1/12/23
---

***Deliverables***
-Upload this R Notebook to your GitHub and submit the link to your Repo on BB.
-Include any graphs or figures created in this assignment in the folder with your R notebook with descriptive file names.

Since this is an optional partner activity, it is okay if your answers are the same as your partnerâ€™s as long as everyone understands it and could explain it in their own words if asked. Each person must individually push their code to Github. *At the top of your R Notebook, write the name of you and your partner(s) as a comment.*

***Complete the following coding activity and answer any following questions as comments in your R Notebook***

In Transcriptomics_Tutorial.Rmd, you learned how to manipulate the SummarizedExperiment data structure, turn it into more readable dataframes, complete basic analysis using those dataframes, and performed a differential expression analysis based on age. Before saving the rna_counts, rna_clinical, and rna_genes however, we masked out any sample that was not a tumor. In this assignment, you will need to reinstantiate the rna_se object and then perform differential expression analysis based on tumor status.

*Pre-Assignment*
Use knitr function to set your working directory to your analysis_data folder in qbio_490_name.
```{r}
 knitr::opts_knit$set(root.dir = normalizePath("/Users/erosmendoza/Downloads/qbio_490_erosmendoza/analysis_data"))
```

Load all necessary packages and instantiate the rna_se object
```{r}
#packages
#BiocManager::install("DESeq2")
#BiocManager::install("EnhancedVolcano")
library(TCGAbiolinks)
library(SummarizedExperiment)
library(DESeq2)
library(EnhancedVolcano)
library(ggplot2)
```

```{r}
#instantiate object
rna_query <- GDCquery(project ="TCGA-SKCM",
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification",
                      workflow.type = "STAR - Counts")

#GDCdownload(rna_query)
rna_se <- GDCprepare(rna_query)
```


*1*
Perform all data cleaning and label manipulation steps (hint: applying colnames and rownames to make the dataframes more readable)

For rna_clinical:
```{r}
#creation:
rna_clinical <- as.data.frame(rna_se@colData)
#removal of "nested lists":
treatments_mask <- colnames(rna_clinical) != "treatments"
rna_clinical <- rna_clinical[, treatments_mask]

primary_site_mask <- colnames(rna_clinical) != "primary_site"
rna_clinical <- rna_clinical[, primary_site_mask]

disease_type_mask <- colnames(rna_clinical) != "disease_type"
rna_clinical <- rna_clinical[, disease_type_mask]

```

For rna_genes:
```{r}
#creation:
rna_genes <- as.data.frame(rna_se@rowRanges@elementMetadata)
#changing rownames
rownames(rna_genes) <- rna_genes$gene_id
```

For rna_counts:
```{r}
#creation:
rna_counts <- as.data.frame(rna_se@assays@data$unstranded)
#changing row names and colnames
rownames(rna_counts) <- rownames(rna_genes)
colnames(rna_counts) <- rownames(rna_clinical)
```


*2*
Determine and define your experimental design for the differential expression analysis, comparing patients samples by gender. Control for covariance of age_category OR vital_status. Manipulate those columns so that they are ready for differential expression analysis (hint: what kind of variables are they? what data type are they by default? do you need to handle unknown values?) Filter out genes with a total expression across all patients less than 10.
```{r}
vital_status_mask <- !is.na(rna_clinical$vital_status)
rna_clinical <- rna_clinical[vital_status_mask, ]

rna_clinical$vital_status <- factor(rna_clinical$vital_status)

design <- ~ gender + vital_status

total_expression <- rowSums(rna_counts)
rna_counts_filtered <- rna_counts[total_expression >= 10, ]

```


*3*
Perform the differential expression analysis, create a volcano plot, and interpret what genes from each of the 6 sections means (assuming thresholds of -log10padj = 2 and log2FoldChange = 1)
```{r}
dds <- DESeqDataSetFromMatrix(countData = rna_counts_filtered,
                              colData = rna_clinical,
                              design= design)

dds_obj <- DESeq(dds)

resultsNames(dds_obj)

results <- results(dds_obj, format = "DataFrame", contrast = c("gender", "female" , "male"))

results <- data.frame(results)
```

Prepare results dataframe for EnhancedVolcano plotting. 
```{r}
#adding information columns and -log10(padj) column
common_genes <- intersect(row.names(results), row.names(rna_genes))

rna_genes_subset <- rna_genes[common_genes, ]

results$gene_name <- rna_genes_subset$gene_name
results$"-log10(padj)" <- -log10(results$padj) 

#row.names(results) <- rna_genes$gene_id
row.names(results) <- row.names(rna_genes_subset)

results
```


```{r}
EnhancedVolcano(results,
                lab = results$gene_name,
                x = 'log2FoldChange',
                y = 'pvalue',
                title = 'Vital Status: Alive vs Dead',
                pointSize = 1.0,
                labSize = 5.0)
ggsave("/Users/erosmendoza/Downloads/qbio_490_erosmendoza/volcano_plot.png", device = "png")
```


# Explain what genes from each part of the Volcano Plot mean in terms of their significance and up/down regulation. 
top-right genes: high significance and large fold change
bottom-right genes: high significance and small fold change
top-left genes: low significance and large fold change
bottom-left genes: low significance and small fold change
top-middle genes: moderate significance and moderate fold change (genes in between top-right and top-left)
bottom-middle genes: moderate significance and moderate fold change (genes in between bottom-right and bottom-left)

Save the picture of the volcano plot (using either ggsave() or right clicking and manually downloading the image and push this .Rmd and the image to GitHub)
